tests/lib/modem_trace/CMakeLists.txt

```diff
@@ -0,0 +1,62 @@
+#
+# Copyright (c) 2021 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+cmake_minimum_required(VERSION 3.20.0)
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(modem_trace_test)
+
+# create mock
+cmock_handle(${ZEPHYR_HAL_NORDIC_MODULE_DIR}/nrfx/drivers/include/nrfx_uarte.h )
+cmock_handle(${ZEPHYR_SEGGER_MODULE_DIR}/SEGGER/SEGGER_RTT.h )
+
+# generate runner for the test
+test_runner_generate(src/modem_trace_test.c)
+
+target_include_directories(app PRIVATE src)
+
+# add test file
+target_sources(app PRIVATE src/modem_trace_test.c)
+
+# add unit under test
+target_sources(app PRIVATE ${NRF_DIR}/lib/nrf_modem_lib/modem_trace.c)
+
+# include paths
+target_include_directories(app PRIVATE ${ZEPHYR_HAL_NORDIC_MODULE_DIR}/nrfx)
+target_include_directories(app PRIVATE ${ZEPHYR_HAL_NORDIC_MODULE_DIR}/nrfx/mdk)
+target_include_directories(app PRIVATE ${ZEPHYR_CMSIS_MODULE_DIR}/CMSIS/Core/Include/)
+target_include_directories(app PRIVATE ${ZEPHYR_SEGGER_MODULE_DIR}/Config/)
+target_include_directories(app PRIVATE ${ZEPHYR_BASE}/modules/hal_nordic/nrfx)
+target_include_directories(app PRIVATE ${NRF_DIR}/include/modem/)
+target_include_directories(app PRIVATE ${ZEPHYR_NRFXLIB_MODULE_DIR}/nrf_modem/include/ )
+
+# Define the device. Otherwise an "unknown device" warning will be triggered when compiling
+# nrfx_config.h
+target_compile_definitions(app PRIVATE NRF9160_XXAA)
+# Define NRFX_DECLARE_ONLY to simplify generation of mocks. If this is not done, then the
+# the compiler would be unable to compile the generated mocks for nrfx uart driver.
+target_compile_definitions(app PRIVATE NRFX_DECLARE_ONLY)
+# Since Device Tree is not available when running unit tests, we need to provide the hardware
+# details that otherwise would be fetched by the UUT from the Device Tree.
+target_compile_definitions(app PRIVATE DT_N_S_uart_1_P_tx_pin=17)
+target_compile_definitions(app PRIVATE DT_N_S_uart_1_P_rx_pin=18)
+
+if(BOARD STREQUAL qemu_cortex_m3)
+    target_compile_definitions(app PRIVATE DT_N_S_soc_S_uart_4000d000=DT_N_S_uart_1)
+endif()
```

Comment:
can you please explain this? why DT isn't available when running unit tests?

Reply:
nicely spotted.
Those should not be needed to pass on as compile definitions, I agree.

Reply:
well. I get an compiler error like this if I dont have it.

`
In file included from /home/balaji/ncs/nrf/tests/lib/modem_trace/src/modem_trace_test.h:11,
                 from <command-line>:
/home/balaji/ncs/nrf/lib/nrf_modem_lib/modem_trace.c: In function 'uart_init':
/home/balaji/ncs/nrf/tests/lib/modem_trace/qmu/zephyr/include/generated/devicetree_unfixed.h:1832:39: error: 'DT_N_S_soc_S_uart_4000d000_P_tx_pin' undeclared (first use in this function); did you mean 'DT_N_S_soc_S_uart_4000d000_P_reg'?
 1832 | #define DT_N_NODELABEL_uart1          DT_N_S_soc_S_uart_4000d000
      |                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/balaji/ncs/zephyr/include/devicetree.h:2641:24: note: in definition of macro 'DT_CAT'
 2641 | #define DT_CAT(a1, a2) a1 ## a2
`

Reply:
Perhaps there is a better way of solving this?

Reply:
The problem seems to be that the code you're trying to test has a hard dependency on nrf uart1 being defined in Devicetree. However, test is expecting to run on native_posix or qemu, platforms where this DT node doesn't exist (and can't exist). So unless the library code is made testable, i.e. properly isolated, what is done here is just wrong.

Reply:
I see. As soon as this PR gets in, we do plan to remove the uart/rtt code from this module and place it in separate modules on their own. Doing that will remove the dependency of UART API from this module.

Reply:
For now, this is a way of mocking the hardware. But yeah, the code will be much easier to test if the UART interfacing code is moved out to a different module, and that will be done as soon as this PR is merged (as mentioned before)

Reply:
Better remove this for now and open a PR with improved code/test (or just add a test that can be run on a platform that has such hardware).

Reply:
Well.. Even if I isolate this module from the device tree specific code, I still need to call another (new) module's API which again interfaces with device tree. Any ideas on how we can test that new module? 
I believe the question boils down to how to unit test a module that has device tree interactions?

Reply:
@gmarull @tejlmand The UART code hasn't changed in this PR. It's only been moved to a different file. @balaji-nordic wrote unit tests (that didn't exist before) using mocks to avoid depending on the actual hardware. So it _is_ possible to run this test on `native_posix` and `qemu`. The tests validate that the unmodified UART implementation is calling the right functions. As far as I understand, the only reason it's needed is because of these two lines in the unmodified code:

    .pseltxd = DT_PROP(DT_NODELABEL(uart1), tx_pin),
    .pselrxd = DT_PROP(DT_NODELABEL(uart1), rx_pin),

And the only place it's being used is when calling the also mocked function `nrfx_uarte_init`. So the only reason the DT definitions are in the test's CMakeLists.txt is to be able to compile the unmodified UART implementation without actually having the device tree. I don't understand why this is so problematic.

Can you please suggest a better way to test the changes without also changing the UART implementation.

Reply:
You may try creating a dt compatible that exposes the same tx/rx_pin properties, and add an overlay for the test that exposes a _fake_ uart1 device.

Reply:
Gave this a shot. Short summary: I could not get that working. Long summary: The [qemu_cortex_m3.dts  ](https://github.com/nrfconnect/sdk-zephyr/blob/main/boards/arm/qemu_cortex_m3/qemu_cortex_m3.dts) includes [lm3s6965.dtsi](https://github.com/nrfconnect/sdk-zephyr/blob/main/dts/arm/ti/lm3s6965.dtsi) which defines a uart1 without `tx-pin` and `rx-pin`. That is the source of the problem. AFAIK, it is not possible to add a new property to a peripheral using an overlay file. (In device tree syntax, there is a `delete-property` but there is no `add-property`). This would mean that we have to create our own .dtsi file and define a uart1 with tx-pin and rx-pin. I tried that but that leads to a compilation error due to the ti-stellaris's console driver unable to get declarations for some peripheral's instance 2 (seems like uart for serial driver). Disabling `CONFIG_SERIAL` makes the compilation succeed (with some warnings) but without SERIAL, nothing gets printed as test output, hence there is no indication of success or failure.

Reply:
To add on: This commit in the zephyr repo fixes the problem completely (no warnings, tests pass)
https://github.com/balaji-nordic/zephyr/commit/5bc81a6e03bf4cd1658436b422928583566ff9ed

Suggestions welcome.

Reply:
@gmarull  Please have a look at the last commit here and see if thats good. I have addressed your comment.